python/ngen_conf/src/ngen/config/path_pair/path_pair.py

```diff
@@ -0,0 +1,277 @@
+    `pathlib.Path`, creating a `PathPairCollection` will return a OS specific subclass
+    (`PosixPathPairCollection` or `WindowsPathPairCollection`).
+
+    The default reader and writer implementations operate on disk. Alternative reader and writer
+    types can be provided assuming they following the `ngen.config.path_path.protocol` `Reader` and
+    `Writer` protocols (interfaces). This could, for example, enable reading and writing from the
+    network.
+
+    serializer's and deserializer's are type specific, and thus have no default. However,
+    `ngen.config.path_path.common` contains generic serializers and deserializer that may be of
+    interest (e.g. `pydantic_serializer` and `pydantic_deserializer`). Alternative serializer and
+    deserializer types can be provided assuming they follow the `ngen.config.path_path.protocol`
+    `Serializer` and `Deserializer` protocols (interfaces).
+    """
+
+    def __new__(
+        cls,
+        *args: StrPath,
+        pattern: str,
+        inner: Optional[List["PosixPathPair[T]"]] = None,
+        reader: Reader = path_reader,
+        writer: Writer = path_writer,
+        serializer: Optional[Serializer[T]] = None,
+        deserializer: Optional[Deserializer[T]] = None,
+        **kwargs: Any,
+    ) -> Union["WindowsPathPairCollection[T]", "PosixPathPairCollection[T]"]:
+        cls = (
+            WindowsPathPairCollection[T]
+            if os.name == "nt"
+            else PosixPathPairCollection[T]
+        )
+
+        template_str = Path(*args)
+
+        if inner is None:
+            inner = []
+
+        prefix, _, suffix = template_str.name.partition(pattern)
+        for item in inner:
+            if PathPairCollection._get_id(item, prefix, suffix) == "":
+                raise ValueError(
+                    f"Filename not derived from template and pattern, {template_str.name!r} {pattern!r}: {item}"
+                )
+
+        self: Union[
+            WindowsPathPairCollection[T], PosixPathPairCollection[T]
+        ] = Path.__new__(
+            cls,
+            *args,
+            inner=inner,
+            reader=reader,
+            writer=writer,
+            serializer=serializer,
+            deserializer=deserializer,
+            **kwargs,
+        )
+        self._inner = inner
+        self._serializer = serializer
+        self._deserializer = deserializer
+        self._reader = reader
+        self._writer = writer
+        self._pattern = pattern
+        return self
+
+    @staticmethod
+    def _get_id(
+        p: "PosixPathPair[T]",
+        prefix: str,
+        suffix: str,
+    ) -> str:
+        assert prefix != "" and suffix != ""
+
+        return p.name[len(prefix) : -len(suffix)]
+
+    @classmethod
+    def cwd(cls) -> Path:
+        return Path.cwd()
+
+    @classmethod
+    def home(cls) -> Path:
+        return Path.home()
+
+    @classmethod
+    def with_objects(
+        cls,
+        o: List[T],
```

Comment:
can we give `o` a little more descriptive name?

Reply:
For sure!